sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,612 @@
+    public Read withMaxNumRecords(long maxNumRecords) {
+      checkArgument(maxReadTime() == null,
+          "maxNumRecord and maxReadTime are exclusive");
+      return builder().setMaxNumRecords(maxNumRecords).build();
+    }
+
+    /**
+     * Define the max read time (duration) while the {@link Read} will receive messages.
+     * When this max read time is not null, the {@link Read} will provide a bounded
+     * {@link PCollection}.
+     */
+    public Read withMaxReadTime(Duration maxReadTime) {
+      checkArgument(maxNumRecords() == Long.MAX_VALUE,
+          "maxNumRecord and maxReadTime are exclusive");
+      return builder().setMaxReadTime(maxReadTime).build();
+    }
+
+    @Override
+    public PCollection<byte[]> apply(PBegin input) {
+
+      org.apache.beam.sdk.io.Read.Unbounded<byte[]> unbounded =
+          org.apache.beam.sdk.io.Read.from(new UnboundedMqttSource(this));
+
+      PTransform<PBegin, PCollection<byte[]>> transform = unbounded;
+
+      if (maxNumRecords() != Long.MAX_VALUE) {
+        transform = unbounded.withMaxNumRecords(maxNumRecords());
+      } else if (maxReadTime() != null) {
+        transform = unbounded.withMaxReadTime(maxReadTime());
+      }
+
+      return input.getPipeline().apply(transform);
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      // validation is performed in the ConnectionConfiguration create()
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      connectionConfiguration().populateDisplayData(builder);
+      if (maxNumRecords() != Long.MAX_VALUE) {
+        builder.add(DisplayData.item("maxNumRecords", maxNumRecords()));
+      }
+      builder.addIfNotNull(DisplayData.item("maxReadTime", maxReadTime()));
+    }
+
+  }
+
+  /**
+   * Checkpoint for an unbounded MQTT source. Consists of the MQTT messages waiting to be
+   * acknowledged and oldest pending message timestamp.
+   */
+  @DefaultCoder(AvroCoder.class)
+  private static class MqttCheckpointMark implements UnboundedSource.CheckpointMark {
+
+    private MqttClient client;
+
+    private final List<MqttMessageWithTimestamp> messages = new ArrayList<>();
+    private Instant oldestPendingTimestamp = BoundedWindow.TIMESTAMP_MIN_VALUE;
+
+    public MqttCheckpointMark() {
+    }
+
+    public void setClient(MqttClient client) {
+      this.client = client;
+    }
+
+    public void add(MqttMessageWithTimestamp message) {
+      if (message.getTimestamp().isBefore(oldestPendingTimestamp)) {
+        oldestPendingTimestamp = message.getTimestamp();
+      }
+      messages.add(message);
+    }
+
+    @Override
+    public void finalizeCheckpoint() {
+      for (MqttMessageWithTimestamp message : messages) {
+        try {
+          client.messageArrivedComplete(message.getMessage().getId(),
+              message.getMessage().getQos());
+          if (message.getTimestamp().isAfter(oldestPendingTimestamp)) {
```

Comment:
The purpose was here to update the watermark when we send the ack to the broker for each message.
However, I think it's overkill. I remove this part.